In [ ]:
# Calculating Factorial

In [ ]:
int fact(int n) {
    
    if (n < 0) {
        throw 1;
    }
    
    int res = 1;
    
    for (int i = 1; i <= n; i++){
        res = res * i;
    }
    
    return res;
}

// TC: O(n)
// SC: O(1)

In [1]:
def fact(n):
    res = 1
    for i in range(1, n+1):
        res = res * i
        
    return res

print(fact(5))

120


**WAP to generage all factorial values stored in an arry till given range**  
5 => [1, 1, 2, 6, 24, 120]

In [ ]:
int fact(int n) {
    
    if (n < 0) {
        throw 1;
    }
    
    int res = 1;
    
    for (int i = 1; i <= n; i++){
        res = res * i;
    }
    
    return res;
}

int[] factList(int n) {
    int res[n+1];
    
    for (i=0; i <=n; i++) {
        res[i] = fact(i);
    }
    
    return res;
}

// TC: O(n^2)
// SC: O(n)

In [ ]:
int[] factList1(int n) {
    int res[n+1];
    
    for (i=0; i <=n; i++) {
        int f = 1;
    
        for (int j = 1; j <= n; j++){
            f = f * j;
        }
        
        res[i] = f;
    }
    
    return res;
}

// TC: O(n^2)


int[] factList2(int n) {
    int res[n+1];
    
    int f = 1;
    for (i=0; i <=n; i++) {
        if (i != 0)
            f = f * i;
        
        res[i] = f;
    }
    
    return res;
}

// TC: O(n)

In [ ]:
for i = 0; i < n ; i ++
    for j = i; j < n; j++
        // some op
        
1 + 2 + 3 + 4... n
=> O(n^2)

In [ ]:
func1():
    for j = i; j < n; j++
        // some op
        
func2():
    for i = 0; i < n ; i ++
        func1()  

func2();
O(n^2)

In [ ]:
int something(int n) {
    count = 0
    while(n > 0) {
        n = n-1;
        count = count + 1;
    }
    
    return count;
}

O(n)

In [ ]:
int something(int n) {
    count = 0
    while(n > 0) {
        n = n/2;
        count = count + 1;
    }
    
    return count;
}

// O(log n)
n=8 : 8 4 2 1 0 -> 4
n=16: 16 8 4 2 1 0 -> 5
n=32: 32 16 8 4 2 1 0 -> 6
n=64: 64 32 16 8 4 2 1 0 -> 7 -> 1 + log(n)

In [2]:

// Loop-1
for (i=1; i < n ; i*=2){
    
}


// Loop-2
for (i=n; i > 0 ; i/=2){
    
}
// O(log n)

SyntaxError: invalid syntax (Temp/ipykernel_10960/187495515.py, line 1)

In [ ]:
for i =0; i < 2**n; i++

O(2^n)

In [ ]:
void moveZerosRight(std::vector<int> array) {
    int nonZeroPos = 0;
    
    int countZeros = 0;
    for(int i =0; i < array.size(); i++) {
        if (array[i] != 0) {
            array[nonZeroPos] = array[i];
            nonZeroPos++;
        } else {
            countZeros++;
        }
    }
    
    while(nonZeroPos < array.size()) {
        array[nonZeroPos] = 0;
        nonZeroPos++;
    }
}

In [ ]:
# Replace every element with the largest element on it's right